From c2e47515b7ca9d6398105888063d1aa0a18508d3 Mon Sep 17 00:00:00 2001
From: Greg Ungerer <gerg@linux-m68k.org>
Date: Tue, 26 Apr 2022 12:14:24 +1000
Subject: [PATCH] uClibc-0.9.33.2: add ELF-FDPIC support for m68knommu

Add support for ELF-FDPIC executables to the m68k startup code.
This is really just ELF support not full FDPIC support.

The relocation (link/loading) is done in a separate uld.so.1 (ELF
interpreter) binary. Just a few tweaks in here to get the stack
arguments right for the slight differences with the ELF-FDPIC loader.

This is a work in progress and used for testing the kernels new m68k
ELF-FDPIC support.

Signed-off-by: Greg Ungerer <gerg@linux-m68k.org>
---
 crt1.S |   68 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++---------
 1 file changed, 59 insertions(+), 9 deletions(-)

diff --git a/libc/sysdeps/linux/m68k/crt1.S b/libc/sysdeps/linux/m68k/crt1.S
--- uClibc/libc/sysdeps/linux/m68k/crt1.S.org	2022-04-24 10:21:31.336553342 +1000
+++ uClibc/libc/sysdeps/linux/m68k/crt1.S	2022-04-26 01:27:51.799747075 +1000
@@ -78,9 +78,13 @@
 	sub.l %fp, %fp
 
 #if !defined __ARCH_USE_MMU__ && defined __PIC__
+#ifdef FDPIC_LOADER
+	move.l #_GLOBAL_OFFSET_TABLE_, %a5
+#else
 	/* Set up the global pointer.  The GOT is at the beginning of the
 	   data segment, whose address is in %d5.  */
 	move.l %d5,%a5
+#endif
 	.equ have_current_got, 1
 #endif
 	
@@ -92,11 +96,11 @@
 	   arguments for `main': argc, argv.  envp will be determined
 	   later in __libc_start_main.  */
 	move.l (%sp)+, %d0	/* Pop the argument count.  */
-#ifndef __ARCH_USE_MMU__
-	move.l (%sp)+, %a0
-#else
+#if defined(__ARCH_USE_MMU__) || defined(FDPIC_LOADER)
 	move.l %sp, %a0		/* The argument vector starts just at the
 				   current stack top.  */
+#else
+	move.l (%sp)+, %a0
 #endif
 
 	/* Provide the highest stack address to the user code (for stacks
-- 
2.25.1
